### DISCUSSION TITLE
FSK, Wave, Packet 그리고 ASK 짧은? 요약

### DISCUSSION THREAD

**Discussion by kuro11pow2** [euphony-io/euphony#115] [Knowledge Share]
# FSK.cpp 분석

코드를 다 적으면 너무너무 길어져서 가독성을 해치기 때문에 설명 위주로 적은 부분 양해 부탁드립니다 ㅠㅠ 😢 
전체 흐름을 먼저 이해하시고 코드랑 같이 보시면 도움이 될 거라고 생각합니다. (되었으면 좋겠네요 ㅠ)
잘못된 부분 있으면 언제든지 댓글로 알려주세요 @designe 👍 

## FSK 사용법
크게 두 가지 mod와 demod 메소드를 사용합니다.
```cpp
auto modulateFSK = fsk->modulate(inputCode);
auto demodulateResult = fsk->demodulate(modulateFSK);
```

메소드의 파라미터와 리턴을 정리하면 다음과 같습니다.

modulate
string -> wave의 벡터
packet -> wave의 벡터

demodulate
wave의 벡터 -> packet
float, 길이, 버퍼크기 -> packet

packet과 wave의 벡터는 mod에도 demod에도 있으니 두 클래스가 중요해보이네요.

## Packet
Packet은 음파 통신의 데이터 패킷을 나타내는 클래스입니다. 통신에서 사용하는 그 패킷 개념이 맞습니다.
PacketBuilder 클래스를 사용하여 Packet을 생성할 때 인자를 직접 넣는 대신 메소드를 호출하여 각 인자를 설정하고 마지막에 build()하여 Packet을 얻습니다.

생성자를 봅시다
```cpp
Packet(const HexVector& source);
Packet(const BaseType type, const HexVector& source);
```

HexVector를 받아서 Base2나 Base16 데이터로 변환하여 payload에 저장합니다.

## Wave
Wave는 디지털 데이터를 아날로그(음파)로 변환한 sin 파형이 저장된 자료구조입니다.
마찬가지로 WaveBuilder를 사용하여 생성합니다.

생성자를 보죠
```cpp
Wave(int hz, int bufferSize);
Wave(const float* src, int bufferSize);
```

주파수와 버퍼 크기를 받거나 float과 버퍼 크기를 받습니다.
주파수와 버퍼 크기를 받은 경우, 멤버 변수를 세팅하고 oscillate를 호출합니다.
oscillate는 뭘하냐? sin 함수를 샘플링해서 버퍼에 채웁니다.
러프하게 설명하자면 0부터 2pi까지 버퍼 크기만큼 등분을 해서 등분한 x 좌표에 대한 sin(x) 값을 저장한다는 것입니다.
물론 주파수에 따라 phase 증가량이 바뀌기 때문에 0부터 2pi까지라는 보장은 없습니다.
그리고 crossfade를 적용하는데 아마도 갑작스럽게 주파수가 바뀔 때 발생하는 잡음을 줄이기 위한 것으로 보입니다.

## FSK
다시 FSK로 돌아와서 modulate를 보겠습니다.
```cpp
WaveList FSK::modulate(Packet* packet)
{
    return this->modulate(packet->toString());
}
```
packet을 문자열로 다시 돌려서 modulate를 합니다. 
packet을 문자열로 돌리면 ```"S" + payload의 문자열 + checksum의 문자열 + parityCode의 문자열```이 됩니다.

```cpp
TestParamType("a", "S6197"),
```
즉 "a"를 Packet으로 만들어서 다시 문자열로 돌린 것이 S + 61 + 9 + 7 인 겁니다. 참고로 a는 0x61입니다.
그럼 이 패킷의 문자열을 modulate에 넣으면 어떻게 될까요.

```cpp
    for (char c : code ) {
        switch(c) {
            case 'S':
                result.push_back(
                        Wave::create()
                                .vibratesAt(kStartSignalFrequency)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
            case '0': case '1': case '2':
            case '3': case '4': case '5':
            case '6': case '7': case '8':
            case '9':
                result.push_back(
                        Wave::create()
                                .vibratesAt(kStandardFrequency + ((c - '0') * kFrequencyInterval))
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
```
이런 식으로 잘 분쇄해서 한 글자씩 Wave로 만들어 벡터에 담아서 return합니다.

다시말해서 FSK가 하는 일은 문자열를 받아서 각 문자마다 정해진 주파수의 wave로 만들어서 돌려주는 것입니다.
그런데 저희가 개발할 ASK는 주파수를 건드리지 않습니다. 그냥 아무 주파수나 사람 귀에 잘 안 들리는 주파수 고정이면 됩니다.
변할 것은 amplitude입니다.  
각 문자마다 amplitude를 정해서 wave의 버퍼에 곱해주면 될 것 같습니다. 문제는 버퍼가 private여서 직접 곱할 수는 없습니다.
WaveBuilder에 amplitude를 설정할 수 있는 메소드도 없습니다.
위에서 혜성님이 말씀하신 것처럼 EuPIOscillator 클래스에 amplitude가 존재는 하지만 이 클래스는 Euphony Programming Interface의 Oscillator 역할을 하는 클래스라 사용하기 어려울 것 같습니다.

제 생각에는 wave와 builder에 amplitude를 설정하는 메소드를 추가하면 쉽게 modulate를 구현할 수 있을 것 같습니다.
그리고 demodulate의 경우 미리 고정해둔 주파수대역의 amplitude를 확인하고 처리해서 HexVector에 차곡차곡 넣은 후에 Packet으로 만들어 return하면 될 것 같습니다.

***

또는 멘토님 말씀처럼 amplitude를 0과 1로만 구분하는 방법도 있을 것 같습니다.
https://github.com/euphony-io/euphony/discussions/85#discussioncomment-1454129

FSK에서는 현재 소리에서 가장 많이 포함된 주파수를 찾고 이를 데이터로 읽지만
ASK에서는 현재 소리의 특정한 주파수의 세기를 구간으로 나눠서 데이터로 읽기 때문에 노이즈에 많이 취약합니다.
그래서 소리의 세기를 잘게 나누면 인식이 잘 안될 것 같습니다.
특정 세기 이하는 0 이상은 1 이런 식으로 BASE2로 전송하면 속도는 많이 느리지만 노이즈에 강해질 것 같네요.

**Comment by judemin** [comment]
어려운 내용 알기 쉽게 정리해주셔서 감사합니다 ㅎㅎ

**Reply by kuro11pow2** [reply]
😄

**Comment by aiclaudev** [comment]
꼼꼼히 읽어봐야겠네요! 감사합니다 👍

**Reply by kuro11pow2** [reply]
😄

**Comment by jopopcorn** [comment]
와 정말 상세하게 설명해주셨네요! ASK에서 주파수는 그냥 기본값으로 처리해야 할 것 같았는데 역시나군요!

> 다시말해서 FSK가 하는 일은 문자열를 받아서 각 문자마다 정해진 주파수의 wave로 만들어서 돌려주는 것입니다.
> 그런데 저희가 개발할 ASK는 주파수를 건드리지 않습니다. 그냥 아무 주파수나 사람 귀에 잘 안 들리는 주파수 고정이면 됩니다.
> 변할 것은 amplitude입니다.
> 각 문자마다 amplitude를 정해서 wave의 버퍼에 곱해주면 될 것 같습니다. 문제는 버퍼가 private여서 직접 곱할 수는 없습니다.
> WaveBuilder에 amplitude를 설정할 수 있는 메소드도 없습니다.
> 위에서 혜성님이 말씀하신 것처럼 EuPIOscillator 클래스에 amplitude가 존재는 하지만 이 클래스는 Euphony Programming Interface의 Oscillator 역할을 하는 클래스라 사용하기 어려울 것 같습니다.
> 
> 제 생각에는 wave와 builder에 amplitude를 설정하는 메소드를 추가하면 쉽게 modulate를 구현할 수 있을 것 같습니다.
> 그리고 demodulate의 경우 미리 고정해둔 주파수대역의 amplitude를 확인하고 처리해서 HexVector에 차곡차곡 넣은 후에 Packet으로 만들어 return하면 될 것 같습니다.

요 마지막 부분을 읽으면서 제가 어떻게 구현해야 할지 고민한 것들이 싹 내려갔습니다.. 멋진 해결책 감사드립니다!! 👍👍
oscillate도 어떤 기능을 수행하는지 와닿지 않았는데 설명 정말 감사드려요! 참고해서 최대한 해결해보도록 하겠습니다!! 💪

**Reply by kuro11pow2** [reply]
감사합니다! 막히는 부분 생기면 공유하면서 함께 만들어봐요!

**Comment by hoho00** [comment]
감사합니다!! 그럼 Amplitude도 0~f, S 까지 해서 16개로 적절하게 나눠서 setting 하는 과정이 필요하겠네요

**Comment by designe** [comment]
아래는 euphony.js 에서의 ASK 구현체입니다. 참고하시면 구현에 도움이 될 것 같습니다.
https://github.com/euphony-io/euphony.js/blob/7441a68df823d06166c92a89f511f043b1c48333/euphony.js#L114

그리고 ASK에서는 하나의 frequency만 사용하시면 되고 (예를 들어 FSK에서의 0에 해당하는 frequency?)
Amplitude는 간단히 말하자면 음파의 세기이자 볼륨입니다. 
거리가 멀어지면 이것이 작아지기 때문에 사실상 볼륨으로 몇단계를 구분하는건 불가능에 가깝습니다.
ASK 조사하셨던 것처럼 해당 frequency의 발생 유무로 0 또는 1로 표현하시면 됩니다.

**Reply by designe** [reply]
ASK를 구현해보시면 데이터 통신의 기초를 직접 구현해보시는 것이기 때문에
통신에 대한 이해도가 부쩍 상승하리라고 봅니다!

**Reply by jopopcorn** [reply]
조언 감사합니다!! setAmplitude를 WaveBuilder에 추가해서 구현해보고 있었는데, 이렇게 하면 이곳저곳 코드를 건드려야해서 이게 맞는건가 싶었습니다ㅠ
그럼 말씀하신 것처럼 0을 표현하려면 Wave를 만들 때 `vibratesAt`을 0을 주고, 1을 표현하려면 기본 주파수값(`kStandardFrequency`)을 주어도 괜찮은가요?

modulate 함수를 다음과 같이 구현해보았는데, 발생 유무로 0과 1을 표현한다는 것이 확실하게 와닿지 않고 제가 잘하고 있는건지 모르겠어서 질문 드립니다!
```c++
WaveList ASK::modulate(std::string code) {
    vector<shared_ptr<Wave>> result;

    for (char c: code) {
        switch (c) {
            case '0':
                result.push_back(
                        Wave::create()
                                .vibratesAt(0)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
            case '1':
                result.push_back(
                        Wave::create()
                                .vibratesAt(kStandardFrequency)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
            case 'S':
                result.push_back(
                        Wave::create()
                                .vibratesAt(kStartSignalFrequency)
                                .setSize(kBufferSize)
                                .setCrossfade(BOTH)
                                .build()
                );
                break;
        }
    }

    return result;
}
```

**Reply by designe** [reply]
매우 매우 좋은 질문입니다!
cpp가 처음이신대로 너무너무 잘해주고 있는 것 같습니다. :+1: 

1)
말씀하신대로 구현하셔도 될 것 같고
왜냐하면 0hz 로 주파수를 발생시키면 아무 것도 발생하지 않을 것이기 때문에 ㅎㅎ
자연스럽게 amplitude가 0 값이 될 것 같구요!

2)
여기 40라인이 ASK 구현부인데 v0.7.1.6에서는 해당 부분에 getZeroSource 라는 함수를 사용해 처리하고 있습니다.
이런식도 가능하구요 :)
https://github.com/euphony-io/euphony/blob/9a72b754133108222abe76c5e0517355b9bb973c/euphony/src/main/java/euphony/lib/transmitter/EuCodeMaker.java#L40

3)
amplitude 가 특별하게 명시된 부분이 없으니 혼란을 줄 수 있는 부분인 것 같습니다.
WaveBuilder 에서
```cpp
Wave::create()
                                .vibratesAt(kStandardFrequency)
                                .setSize(kBufferSize)
                                .setAmplitude(0.0) // 새로 구현해야 할 함수
                                .build()
```
위와 같이 접근해도 될 것 같구요!
 이 방법이 programmable한 좋은 방법입니다만 추가적으로 입력받은 amplitude 0.0 ~ 1.0을 
https://github.com/euphony-io/euphony/blob/ee76f7890caf24848d09a3109b7b3be745d82ac5/euphony/src/main/cpp/core/source/Wave.cpp#L58
여기 라인에서 sin함수 에 곱하기해주는 작업까지 필요합니다 :)

4)
```cpp
Wave::create()
                                .vibratesAt(kStandardFrequency)
                                .setSize(kBufferSize)
                                .inMuted() // 새로 구현
                                .build()
```
이것이 가장 간단한 방법일 것 같은데 inMuted
inMuted() 함수가 해야할 일은 그냥 0 값으로 이루어진 buffer 만들어 리턴해주는 것입니다!
2번과 똑같은 방법이라고 볼 수 있는데 WaveBuilder 에서 대신 처리해주는 것이죠.

4번이 가장 빨리 리펙토링할 수 있을 것 같네요.
현재 방법도 문제없어보입니다 :)

**Reply by jopopcorn** [reply]
답변 정말정말 감사드립니다!! 여러가지 구현 방법을 설명해주셔서 이해하는데 도움이 많이 됐습니다!
amplitude가 명시되지 않아 혼란을 줄 수 있다는 점 공감됩니다..ㅎㅎ 저부분을 어떻게 처리해야할지 며칠동안 고민을 많이 했는데 잘하고 있었다니 정말 다행이네요!
나중에 3번 아니면 4번으로 리팩토링 해보면 좋을거같고, 우선은 시간이 촉박하니 1번으로 계속 진행해보도록 하겠습니다!!

**Reply by kuro11pow2** [reply]
@designe 
멘토님 .build() 에서 osillator를 호출하고 있어서 .inMuted()에서 wave의 mSource를 수정해도 무시당할 것 같습니다! 

WaveBuilder.cpp
```cpp
std::shared_ptr<Wave> WaveBuilder::build() {
    if(wave.getSize() > 0 && wave.getHz() > 0) {
        wave.oscillate();
    }

    return std::make_shared<Wave>(wave);
}
```

**Reply by designe** [reply]
> @designe 멘토님 .build() 에서 osillator를 호출하고 있어서 .inMuted()에서 wave의 mSource를 수정해도 무시당할 것 같습니다!
> 
> WaveBuilder.cpp
> 
> ```c++
> std::shared_ptr<Wave> WaveBuilder::build() {
>     if(wave.getSize() > 0 && wave.getHz() > 0) {
>         wave.oscillate();
>     }
> 
>     return std::make_shared<Wave>(wave);
> }
> ```
@kuro11pow2 
앗 웨이브쪽은 무조건 건들수밖에 없는 구조겠군요. ㅎㅎ
만약 inMuted()함수로 간다면 wave쪽 수정과 함께 3번과 유사하게 가야하겠네요!
좋은 코멘트입니다!

**Comment by dlwlstks96** [comment]
자세한 설명 감사드립니다!!!🔥🔥
